### ISSUE TITLE
How to determine action for different key release within a button?

### ISSUE THREAD

**Issue by obar** [david-janssen/kmonad#58]
I've started learning Haskell so I could do something more useful in this project than just asking for things :) I'm beginning to get it, but I'm not sure how to capture a KeyEvent and determine the action from it. Or maybe I am thinking about this problem the wrong way.

What I'd like to do is have a `tapNextRelease` parameter to treat modifiers of the tap action as sticky if they were down at the press of this button. An example:

```
sc (tap-next-release s lctl)
```

If tapped or in a roll it emits `s`, but only on key release—that's not how a key normally works with shift for example, emitting on key press. `s` is only capitalized when shift is held until release, which feels quite different than the other keys. Rather than slowing down all typing that involves the shift key, what I'd like is to capture a release of shift during the press and apply it during the tap action. That's not too hard, but the complexity arises when using different keys that emit these modifiers. Continuing the example:

```
/sh (tap-hold-next 300 / rsft)
```

The KeyEvent for this key is `KeySlash`, not a shift. Is there a way to capture the translated KeyEvent from this? Here is the code I have so far

```haskell
myTapNextRelease :: Button -> Button -> Button
myTapNextRelease t h = onPress $ do
  hold True
  go [] []
  where
    allmods = [KeyLeftCtrl,KeyLeftShift,
               KeyLeftAlt,KeyLeftMeta,
               KeyRightCtrl,KeyRightShift,
               KeyRightAlt,KeyRightMeta,
               KeyCompose, KeySlash, KeyCapsLock]
    -- What it should actually be:
    -- allmods = [KeyLeftCtrl,KeyLeftShift,
    --            KeyLeftAlt,KeyLeftMeta,
    --            KeyRightCtrl,KeyRightShift,
    --            KeyRightAlt,KeyRightMeta,
    --            KeyCompose]
    go :: MonadK m => [Keycode] -> [Keycode] ->  m ()
    go ks mods = hookF $ \e -> do
      p <- matchMy Release
      let isRel = isRelease e
      if
        -- If the next event is my own release: we act as if we were tapped
        | p e -> doTap mods
        -- If the next event is the release of some button that was held after me
        -- we act as if we were held
        | isRel && (e^.keycode `elem` ks) -> doHold e
        -- If the next event is the release of a modifier, store that modifier and wait again
        | isRel && (e^.keycode `elem` allmods) -> (go ks ((e^.keycode):mods)) *> pure NoCatch
        -- Else, if it is a press, store the keycode and wait again
        | not isRel                       -> go ((e^.keycode):ks) mods *> pure NoCatch
        -- Else, if it is a release of some button held before me, just ignore
        | otherwise                       -> go ks mods *> pure NoCatch

    -- Behave like a tap: tap the button `t`, mod as needed, release processing
    doTap :: MonadK m => [Keycode] -> m Catch
    doTap [] = tap t *> hold False *> pure Catch
    doTap mods =
      case (head mods) of
        KeySlash    -> tap (modded KeyRightShift t) *> hold False *> pure Catch
        KeyCapsLock -> tap (modded KeyLeftCtrl t) *> hold False *> pure Catch
        _           -> tap (modded (head mods) t) *> hold False *> pure Catch

    -- Behave like a hold is not simple: first we release the processing hold,
    -- then we catch the release of ButtonX that triggered this action, and then
    -- we rethrow this release.
    doHold :: MonadK m => KeyEvent -> m Catch
    doHold e = press h *> hold False *> inject e *> pure Catch
```

(note that this requires importing `Data.List` and adding `myTapNextRelease` to Joiner.hs, Parser.hs and Types.hs)

For now to keep things easy for me, I'm only applying one released modifier (for my usage it's sufficient but it should do some sort of fold applying all mods from the list), hard-coding in the slash mod as well as my capslock-as-ctrl translation, and creating a new type of button (rather than an argument for the existing button). That stuff I can certainly change later, and as it stands this all works fine for me, but it's clearly in no shape to contribute if special mods need to be hard-coded in. Is it even possible to, in this function, figure out that the release of `keySlash` would emit right shift?

**Comment by david-janssen** [comment]
Hey @obar,

Great news you're trying to learn Haskell. The initial weirdness can be a bit high, but once you adapt it's a great language. Didn't manage to get around answering you this weekend, but I haven't forgotten about you. Will try to find some time tomorrow afternoon.

Cheers,

D

**Comment by david-janssen** [comment]
Hey @obar,

So, I took the time to reread your post a few times to get a feeling for what you wanted, and here's the bad news: Impossible given the current state of the code. Let me go through you post and highlight where I think you might have a misconception, then explain why this is impossible, and finally highlight how we might make this work:

### The misconception
> `/sh (tap-hold-next 300 / rsft)`
> The KeyEvent for this key is KeySlash, not a shift. 

None of the code you quoted is in any way a `KeyEvent`. A `KeyEvent` is exactly that: an event communicated from the OS to kmonad indicated that 'something has happened', or an event that can be communicated back to the OS to indicate that 'something should happen'.

The `(tap-hold-next 300 / rsft)` consists of (sort of) 3 component:
- `/`: This is a 'Button' in KMonad. A `Button` consists of 2 `Action`s (either or both of which can be 'do nothing'), 1 for 'pressing', and 1 for 'releasing'. This specific button emits the press of '/' when pressed, and emits the release of '/' when released.
- `rsft`: This is exactly the same as `/`, with the singular exception of the `Keycode` that gets emitted.
- `(tap-hold-next 300 a b)`: `tap-hold-next` is what I refer to as a button combinator, it takes 1 or more buttons (here 2) and perhaps some parameters (here 300ms) and returns a new button. 

Putting it together, the entire statement: `(tap-hold-next 300 / rsft)` evaluates back to a single `Button`. (on which you could, if you wanted, use more combinators). Once the parsing of the config-file is complete, all of these buttons are the same: first-class citizens in KMonad.

### The issue
Why won't what you want to do work? Well, it's because there is no way to hook output events. You already correctly identified that, although you could try catching input-event `shift` codes, that is no guarantee, because `shift` could just as easily be emitted by some arbitrary button bound to the input event corresponding to `KeyX`. The 'emitting' of `KeyEvent`s in KMonad is currently very simple: we write directly to the output sink, which writes directly to the OS. There is no intermediate machinery.

### A potential solution
So I guess people are starting to come up with scenarios where you want to do different things based on some form of 'global state'. I designed KMonad with a particular principle in mind: Buttons should be independent from eachother and the global state. That is to say, a button should be happy existing as the only button, or one of millions. Buttons should never know about eachother or eachothers internal state. I think allowing for that creates the potential to make a huuuuuge swamp of confusion.

*however*

Thinking your problem through brought me face to face with a very elegant (and simple) solution. We could simply allow for hooking output events. This way, you could start monitoring output events the moment you get pressed, if you detect the emission of the release of a modifier and you decide you are *also* tapping, you could then apply that modifier as you desired. All the hooks (input and output) that come from a single button can share state, so you could even make the hooks play nicely together. Sure, you'd have a bit of headache getting your button to work well in all circumstances, but the confusion of your button would be isolated to just that single button, without leaking into the rest of the implementation.

So: I've put it on top of my pile, I'll make a new issue and reference this thread. Once the feature exists it should be theoretically possible, at the moment, it isn't. :-)

**Comment by obar** [comment]
Very cool, thanks David!

My hacked version of this (has been working quite well for me, but I suspect HookLocation will let a few interesting key behaviors be generalized and used. And since that's the answer to this question I should close this issue :)

**Comment by david-janssen** [comment]
Very good. I am currently wrapping up the documentation for the migration of refactor to master. After that, I intend to seriously redo the hooks, I have some ideas that would probably make a lot of the code more straightforward, and make ideas like yours much easier to implement, but that's something for the future.
